gprofiler/perf.py

```diff
@@ -41,42 +46,43 @@ def __enter__(self):
-        parsed_path = os.path.join(self._storage_dir, f"{filename_base}.parsed")
-
+    def _run_perf(self, dwarf: bool = False) -> str:
         buildid_args = ["--buildid-dir", PERF_BUILDID_DIR]
 
         with NamedTemporaryFile(dir=self._storage_dir) as record_file:
             args = ["-F", str(self._frequency), "-a", "-g", "-o", record_file.name]
             if dwarf:
-                args += ["--call-graph", "dwarf"]
+                args += ["--call-graph", f"dwarf,{self._dwarf_stack_size}"]
             run_process(
                 [resource_path("perf")] + buildid_args + ["record"] + args + ["--", "sleep", str(self._duration)],
                 stop_event=self._stop_event,
             )
-            with open(parsed_path, "w") as f:
-                run_process(
-                    [resource_path("perf")] + buildid_args + ["script", "-F", "+pid", "-i", record_file.name], stdout=f
-                )
-            return parsed_path
+            perf_script_result = run_process(
+                [resource_path("perf")] + buildid_args + ["script", "-F", "+pid", "-i", record_file.name],
+                suppress_log=True,
+            )
+            return perf_script_result.stdout.decode('utf8')
 
-    def snapshot(self) -> Iterable[Mapping[str, str]]:
+    def snapshot(self) -> Tuple[ProcessToStackSampleCounters, ProcessIdToCommMapping]:
         free_disk = psutil.disk_usage(self._storage_dir).free
         if free_disk < 4 * 1024 * 1024:
             raise Exception(f"Free disk space: {free_disk}kb. Skipping perf!")
 
         logger.info("Running global perf...")
-        record_path = self._run_perf("global")
-        logger.info("Finished running global perf")
-        return parse_perf_script(Path(record_path).read_text())
+        if not self._fp_perf:
+            return merge_global_perfs(None, self._run_perf(dwarf=True))
+        if not self._dwarf_perf:
+            return merge_global_perfs(self._run_perf(dwarf=False), None)
+        with concurrent.futures.ThreadPoolExecutor(max_workers=2) as executor:
+            # We are running 2 perfs in parallel - one with DWARF and one with FP, and then we merge their results.
+            # This improves the results from software that is compiled without frame pointers,
+            # like some native software. DWARF by itself is not good enough, as it has issues with unwinding some
+            # versions of Go processes.
+            fp_future = executor.submit(self._run_perf, False)
+            dwarf_future = executor.submit(self._run_perf, True)
+        fp_perf = fp_future.result()
+        dwarf_perf = dwarf_future.result()
 
-        # TODO: run dwarf in parallel, after supporting it in merge.py
-        # Alternatively: fix golang dwarf problems and the run just dwarf
-
-        # if get_cpu_count() > 32:
-        #     dwarf_frequency = 99
-        # else:
-        #     dwarf_frequency = 999
+        logger.info("Finished running global perf")
```

Comment:
This message is not shown anymore if we're not in `auto` mode.

How about you restructure this function so there's only 1 call to `merge_global_perfs` with possible `None` placeholders?